### PR TITLE
[Merged by Bors] - Split vm/opcode into modules

### PR DESCRIPTION
**Pull request by nekevss**
Hi!

This isn't really related to a pull request that I know of. I was trying to better wrap my head around Boa's VM and thought I'd break it apart so that the file wasn't 2500+ lines. I figured I'd submit it as a draft and get feedback/see if anyone was interested in it. The way the modules were broken apart was primarily based off the opcode name (`GetFunction` & `GetFunctionAsync` -> `./get/function.rs`).

It changes the following:

- Adds an `Operation` trait to opcode/mod.rs
- Implements `Operation` for each Opcode variant, moving the executable instruction code from `vm/mod.rs` to the respective module

### CODE REVIEW

**Comment by Razican** [comment]
Great work!! Thank you for this contribution :)

This is related to #1808, but for the VM instead of the byte compiler, if I'm not mistaken.

**Comment by nekevss** [comment]
Just wanted to follow up that this is probably ready for review (updated to some of the recent pull requests). 😄

File: boa_engine/src/vm/opcode/mod.rs [hunk 1]
```diff
@@ -1,4 +1,77 @@
...
+pub(crate) use unary_ops::*;
+pub(crate) use value::*;
+pub(crate) use void::*;
+
+pub(crate) trait Operation {
+    const NAME: &'static str;
+    const INSTRUCTION: &'static str;
+
+    fn execute(context: &mut Context) -> JsResult<ShouldExit>;
+}
```
Comment (Razican) [inline boa_engine/src/vm/opcode/mod.rs:73]: This trait should have some documentation, to clearly explain what does it do, how would someone implement it for a new opcode and so on.

Ideally, all implementing opcodes should have module-level and structure-level documentation, but we are lacking documentation in general for the VM (look at the `Opcode` enum just below). If you feel that documenting all op codes is a bit too much for this PR, let me know, and I'll open a new issue (similar to #1608) to document the VM.

**Comment by Razican** [comment]
### Test262 conformance changes

  #### VM implementation

  | Test result | main count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 93,383 | 93,383 | 0 |
  | Passed | 69,095 | 69,095 | 0 |
  | Ignored | 18,182 | 18,182 | 0 |
  | Failed | 6,106 | 6,106 | 0 |
  | Panics | 0 | 0 | 0 |
  | Conformance | 73.99% | 73.99% | 0.00% |

**Review by Razican — COMMENTED** [review]
Great work!! This is very helpful.

I added a couple of comments. I like the approach, and I would be OK to merge it as-is, but check if my comments make sense. Also, this is missing documentation of all the op-codes, but we didn't have such documentation before, so, if you add it, that would be great, but if not, I'm OK to merge it as-is and create an issue to do it later :)

File: boa_engine/src/vm/opcode/mod.rs [hunk 2]
```diff
@@ -1,4 +1,77 @@
...
+pub(crate) use return_stm::*;
+pub(crate) use set::*;
+pub(crate) use swap::*;
+pub(crate) use switch::*;
+pub(crate) use throw::*;
+pub(crate) use to::*;
+pub(crate) use try_catch::*;
+pub(crate) use unary_ops::*;
+pub(crate) use value::*;
+pub(crate) use void::*;
```
[the remark this answers is not in the document]
Comment (nekevss) [reply boa_engine/src/vm/opcode/mod.rs]: Completed! 😄

**Review by Razican — APPROVED** [review]
Thanks! Looks good to me :)

**Comment by raskad** [comment]
I like the change. I pushed an empty commit to trigger the benchmarks to see if there is any noticeable change in how the code is compiled/optimized.

**Review by RageKnify — APPROVED** [review]

**Review by HalidOdat — APPROVED** [review]
Looks good to me :)

**Comment by Razican** [comment]
Too much standard deviation, difficult to say, but I don't see it went worse, and we might even have some gains.

```
┌─────────┬──────────────────────────────────────────────┬───────────────────────┬───────────────────────┬────────────┐
│ (index) │                     name                     │     baseDuration      │    changesDuration    │ difference │
├─────────┼──────────────────────────────────────────────┼───────────────────────┼───────────────────────┼────────────┤
│    0    │      'Arithmetic operations (Compiler)'      │    '666.2±47.77ns'    │    '667.7±58.22ns'    │   '0.0'    │
│    1    │     'Arithmetic operations (Execution)'      │    '689.4±72.20ns'    │  '**584.2±48.15ns**'  │   '-15'    │
│    2    │       'Arithmetic operations (Parser)'       │   '**6.9±0.45µs**'    │     '7.4±0.53µs'      │   '+7.0'   │
│    3    │          'Array access (Compiler)'           │   '1792.3±95.59ns'    │ '**1778.5±140.42ns**' │  '-0.99'   │
│    4    │          'Array access (Execution)'          │   '**10.8±0.82µs**'   │     '11.2±0.87µs'     │   '+3.0'   │
│    5    │           'Array access (Parser)'            │     '15.0±0.91µs'     │   '**14.8±1.00µs**'   │  '-0.99'   │
│    6    │         'Array creation (Compiler)'          │   '**2.7±0.18µs**'    │     '2.8±0.27µs'      │   '+4.0'   │
│    7    │         'Array creation (Execution)'         │   '1453.8±104.58µs'   │ '**1434.1±101.41µs**' │  '-0.99'   │
│    8    │          'Array creation (Parser)'           │     '19.0±2.55µs'     │   '**18.5±1.56µs**'   │   '-2.9'   │
│    9    │            'Array pop (Compiler)'            │   '**4.6±0.27µs**'    │     '4.9±0.45µs'      │   '+5.0'   │
│   10    │           'Array pop (Execution)'            │  '**866.5±88.80µs**'  │    '872.0±86.13µs'    │   '+1.0'   │
│   11    │             'Array pop (Parser)'             │  '**163.8±10.15µs**'  │    '165.0±8.84µs'     │   '+1.0'   │
│   12    │      'Boolean Object Access (Compiler)'      │ '**1367.0±272.73ns**' │   '1397.8±162.75ns'   │   '+2.0'   │
│   13    │     'Boolean Object Access (Execution)'      │   '**6.5±0.55µs**'    │     '6.6±0.78µs'      │   '+2.0'   │
│   14    │       'Boolean Object Access (Parser)'       │     '17.6±1.46µs'     │   '**17.2±0.97µs**'   │   '-2.9'   │
│   15    │            'Clean js (Compiler)'             │     '6.0±0.33µs'      │     '5.9±0.53µs'      │   '0.0'    │
│   16    │            'Clean js (Execution)'            │    '866.6±53.33µs'    │  '**850.6±53.68µs**'  │   '-2.0'   │
│   17    │             'Clean js (Parser)'              │   '**37.1±3.29µs**'   │     '38.8±5.45µs'     │   '+4.0'   │
│   18    │                'Create Realm'                │  '**284.0±14.95ns**'  │    '312.6±25.01ns'    │   '+10'    │
│   19    │ 'Dynamic Object Property Access (Compiler)'  │     '2.1±0.11µs'      │     '2.1±0.11µs'      │   '0.0'    │
│   20    │ 'Dynamic Object Property Access (Execution)' │     '7.0±0.82µs'      │   '**6.6±0.64µs**'    │   '-4.8'   │
│   21    │  'Dynamic Object Property Access (Parser)'   │     '13.8±1.09µs'     │     '13.8±1.41µs'     │   '0.0'    │
│   22    │            'Fibonacci (Compiler)'            │   '**3.3±0.32µs**'    │     '3.3±0.24µs'      │   '+3.0'   │
│   23    │           'Fibonacci (Execution)'            │ '**1374.5±138.62µs**' │   '1419.7±93.57µs'    │   '+3.0'   │
│   24    │             'Fibonacci (Parser)'             │     '21.1±1.38µs'     │   '**20.7±1.94µs**'   │   '-2.0'   │
│   25    │            'For loop (Compiler)'             │   '**3.2±0.31µs**'    │     '3.3±0.43µs'      │   '+6.0'   │
│   26    │            'For loop (Execution)'            │   '**20.7±1.94µs**'   │     '21.0±1.74µs'     │   '+1.0'   │
│   27    │             'For loop (Parser)'              │     '18.3±1.12µs'     │   '**18.2±0.95µs**'   │  '-0.99'   │
│   28    │             'Mini js (Compiler)'             │   '**5.1±0.47µs**'    │     '5.7±0.92µs'      │   '+13'    │
│   29    │            'Mini js (Execution)'             │    '780.6±45.59µs'    │  '**770.4±44.45µs**'  │  '-0.99'   │
│   30    │              'Mini js (Parser)'              │   '**32.1±1.53µs**'   │     '32.4±3.78µs'     │   '+1.0'   │
│   31    │      'Number Object Access (Compiler)'       │ '**1273.5±394.98ns**' │   '1292.6±115.16ns'   │   '+2.0'   │
│   32    │      'Number Object Access (Execution)'      │     '5.2±0.36µs'      │   '**5.1±0.45µs**'    │   '-2.0'   │
│   33    │       'Number Object Access (Parser)'        │   '**14.0±2.28µs**'   │     '14.1±1.50µs'     │   '+1.0'   │
│   34    │         'Object Creation (Compiler)'         │ '**1908.4±189.70ns**' │     '2.0±0.29µs'      │   '+7.0'   │
│   35    │        'Object Creation (Execution)'         │   '**6.2±0.47µs**'    │     '6.3±0.39µs'      │   '+2.0'   │
│   36    │          'Object Creation (Parser)'          │   '**12.0±0.96µs**'   │     '12.2±0.93µs'     │   '+1.0'   │
│   37    │             'RegExp (Compiler)'              │     '2.1±0.17µs'      │     '2.1±0.16µs'      │   '0.0'    │
│   38    │             'RegExp (Execution)'             │   '**17.1±1.87µs**'   │     '18.8±9.02µs'     │   '+10'    │
│   39    │              'RegExp (Parser)'               │     '13.1±1.39µs'     │   '**12.9±0.88µs**'   │  '-0.99'   │
│   40    │         'RegExp Creation (Compiler)'         │ '**1974.3±163.36ns**' │     '2.0±0.34µs'      │   '+3.0'   │
│   41    │        'RegExp Creation (Execution)'         │     '11.9±2.11µs'     │   '**11.6±0.74µs**'   │   '-2.9'   │
│   42    │          'RegExp Creation (Parser)'          │   '**10.9±0.47µs**'   │     '10.9±1.03µs'     │   '+1.0'   │
│   43    │         'RegExp Literal (Compiler)'          │     '2.1±0.12µs'      │     '2.1±0.14µs'      │   '0.0'    │
│   44    │         'RegExp Literal (Execution)'         │   '**16.9±1.10µs**'   │     '17.7±1.59µs'     │   '+5.0'   │
│   45    │          'RegExp Literal (Parser)'           │     '10.4±0.50µs'     │   '**10.3±0.54µs**'   │  '-0.99'   │
│   46    │     'RegExp Literal Creation (Compiler)'     │ '**1910.2±96.31ns**'  │   '1956.3±117.89ns'   │   '+2.0'   │
│   47    │    'RegExp Literal Creation (Execution)'     │     '11.6±0.94µs'     │     '11.6±0.78µs'     │   '0.0'    │
│   48    │      'RegExp Literal Creation (Parser)'      │     '8.1±0.30µs'      │   '**8.1±0.63µs**'    │  '-0.99'   │
│   49    │  'Static Object Property Access (Compiler)'  │ '**1915.7±153.87ns**' │   '1935.9±127.46ns'   │   '+1.0'   │
│   50    │ 'Static Object Property Access (Execution)'  │     '6.7±0.76µs'      │   '**6.2±0.35µs**'    │   '-8.3'   │
│   51    │   'Static Object Property Access (Parser)'   │   '**12.8±0.72µs**'   │     '12.8±1.50µs'     │   '+1.0'   │
│   52    │      'String Object Access (Compiler)'       │ '**1699.2±105.07ns**' │   '1740.0±139.72ns'   │   '+2.0'   │
│   53    │      'String Object Access (Execution)'      │     '9.3±1.09µs'      │   '**9.1±0.90µs**'    │   '-2.0'   │
│   54    │       'String Object Access (Parser)'        │   '**16.8±1.62µs**'   │     '18.0±1.50µs'     │   '+7.0'   │
│   55    │        'String comparison (Compiler)'        │   '**2.7±0.13µs**'    │     '2.8±0.18µs'      │   '+1.0'   │
│   56    │       'String comparison (Execution)'        │     '5.5±0.53µs'      │   '**5.4±0.31µs**'    │  '-0.99'   │
│   57    │         'String comparison (Parser)'         │     '14.6±3.45µs'     │   '**14.4±2.06µs**'   │  '-0.99'   │
│   58    │      'String concatenation (Compiler)'       │   '**2.2±0.17µs**'    │     '2.2±0.21µs'      │   '+1.0'   │
│   59    │      'String concatenation (Execution)'      │   '**5.0±0.41µs**'    │     '5.2±0.52µs'      │   '+3.0'   │
│   60    │       'String concatenation (Parser)'        │     '10.0±0.98µs'     │   '**9.7±0.55µs**'    │   '-2.0'   │
│   61    │           'String copy (Compiler)'           │ '**1769.8±99.02ns**'  │   '1800.9±114.46ns'   │   '+2.0'   │
│   62    │          'String copy (Execution)'           │     '5.0±0.61µs'      │   '**4.9±0.29µs**'    │   '-2.0'   │
│   63    │            'String copy (Parser)'            │   '**7.4±0.54µs**'    │     '7.5±0.71µs'      │   '+1.0'   │
│   64    │             'Symbols (Compiler)'             │   '1375.1±138.63ns'   │ '**1366.5±121.95ns**' │  '-0.99'   │
│   65    │            'Symbols (Execution)'             │     '5.3±0.62µs'      │   '**4.9±0.34µs**'    │   '-7.4'   │
│   66    │              'Symbols (Parser)'              │     '5.9±0.60µs'      │     '5.9±0.45µs'      │   '0.0'    │
│   67    │                      ''                      │       undefined       │       undefined       │   '+NaN'   │
└─────────┴──────────────────────────────────────────────┴───────────────────────┴───────────────────────┴────────────┘
```

**Review by raskad — APPROVED** [review]

**Pull request closed without merging**